src/Microsoft.AspNet.Http.Abstractions/Authentication/AuthenticationManager.cs

```diff
@@ -11,41 +13,66 @@ public abstract class AuthenticationManager
         {
-            Challenge(authenticationScheme: null, properties: properties);
+            return ChallengeAsync(authenticationScheme: null, properties: properties);
         }
 
-        public virtual void Challenge(string authenticationScheme)
+        public virtual Task ChallengeAsync([NotNull] string authenticationScheme)
         {
-            Challenge(authenticationScheme: authenticationScheme, properties: null);
+            return ChallengeAsync(authenticationScheme: authenticationScheme, properties: null);
         }
 
-        public abstract void Challenge(string authenticationScheme, AuthenticationProperties properties);
+        // Leave it up to authentication handler to do the right thing for the challenge
+        public Task ChallengeAsync([NotNull] string authenticationScheme, AuthenticationProperties properties)
+        {
+            return ChallengeAsync(authenticationScheme, properties, ChallengeBehavior.Automatic);
+        }
 
-        public void SignIn(string authenticationScheme, ClaimsPrincipal principal)
+        public Task SignInAsync([NotNull] string authenticationScheme, ClaimsPrincipal principal)
         {
-            SignIn(authenticationScheme, principal, properties: null);
+            return SignInAsync(authenticationScheme, principal, properties: null);
         }
 
-        public abstract void SignIn(string authenticationScheme, ClaimsPrincipal principal, AuthenticationProperties properties);
+        public Task ForbidAsync([NotNull] string authenticationScheme)
+        {
+            return ForbidAsync(authenticationScheme, properties: null);
+        }
 
-        public virtual void SignOut()
+        // Deny access (typically a 403)
+        public Task ForbidAsync([NotNull] string authenticationScheme, AuthenticationProperties properties)
         {
-            SignOut(authenticationScheme: null, properties: null);
+            return ChallengeAsync(authenticationScheme, properties, ChallengeBehavior.Forbidden);
         }
 
-        public abstract void SignOut(string authenticationScheme);
+        public abstract Task ChallengeAsync([NotNull] string authenticationScheme, AuthenticationProperties properties, ChallengeBehavior behavior);
+
+        public abstract Task SignInAsync([NotNull] string authenticationScheme, ClaimsPrincipal principal, AuthenticationProperties properties);
+
+        public Task SignOutAsync()
+        {
+            return SignOutAsync(authenticationScheme: null, properties: null);
+        }
+
+        public Task SignOutAsync([NotNull] string authenticationScheme)
+        {
+            return SignOutAsync(authenticationScheme, properties: null);
+        }
 
-        public abstract void SignOut(string authenticationScheme, AuthenticationProperties properties);
+        public abstract Task SignOutAsync([NotNull] string authenticationScheme, AuthenticationProperties properties);
```

Comment:
Out of curiosity, what are you expecting [NotNull] to do on an abstract method? I found EF doing the same thing recently.